File: Csharp/Lab8/Lab8/Lab8/cng/Model/Collections/neww/stack/StackImpl.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Lab8.cng.Model.Collections.neww.stack
{
    public class StackImpl<T> : AbstractCollection<T>, IStackInterface<T>
    {
       
        private object[] _stack = new object[4];
        private int _sizeVar;
        private int _stackArraySize = 4;

        public StackImpl()
        {
        }

        public StackImpl(T[] ts)
        {
            foreach (T t in ts)
            {
                Push(t);
            }
        }


        public override int Size()
        {
            return _sizeVar;
        }

        public override void Add(T t)
        {
            Push(t);
        }

        public void Push(T t)
        {
            if (_sizeVar == _stackArraySize - 1)
            {
                Resize(_stackArraySize * 2);
            }

            _stack[_sizeVar++] = t;
        }


        public T Peek()
        {
            return (T) _stack[_sizeVar - 1];
        }


        public T Pop()
        {
            T t = (T) _stack[_sizeVar - 1];
            _stack[_sizeVar--] = null;
            return t;
        }


        //iterator
        public override IEnumerator<T> GetEnumerator()
        {
            for (int i = 0; i < _sizeVar; i++)
            {
                yield return (T) _stack[i];
            }
        }
        //iterator


        public override int GetHashCode()
        {
            const int hashconst = 31;
            int result = 1;
            foreach (object o in this._stack)
            {
                result += result * hashconst + o.GetHashCode();
            }

            return result;
        }

       
        public bool Equals(object obj)
        {
            return this.GetHashCode() == obj.GetHashCode() && obj.GetType() == this.GetType();
        }


        public string ToString()
        {
            StringBuilder sb = new StringBuilder().Append("Content\n");
            for (int i = 0; i < _sizeVar; i++)
            {
                sb.Append(i).Append("\n");
            }

            return sb.ToString();
        }

        
        //PRIVATE METHODS

        private void Resize(int newLength)
        {
            _stackArraySize = newLength;
            Object[] newArray = new Object[newLength];
            for (int i = 0; i < _sizeVar; i++)
            {
                newArray[i] = _stack[i];
            }

            _stack = newArray;
        }


        
    }
}

File: Java/Desktop/NewLab8/src/by/bntu/fitr/povt/cng/model/collections/neww/stack/StackImpl.java
package by.bntu.fitr.povt.cng.model.collections.neww.stack;

import by.bntu.fitr.povt.cng.model.collections.neww.AbstractCollection;

import java.util.Iterator;

public class StackImpl<T> extends AbstractCollection<T> implements StackInterface<T> {

    private Object[] stack = new Object[4];
    private int sizeVar;
    private int stackArraySize = 4;

    public StackImpl(){
    }

    public StackImpl(T...ts){
        for (T t : ts) {
            push(t);
        }
    }

    @Override
    public int size() {
        return sizeVar;
    }

    @Override
    protected void add(T t) {
        push(t);
    }

    @Override
    public T peek() {
        return (T) stack[sizeVar - 1];
    }

    @Override
    public void push(T t) {
        if (sizeVar == stackArraySize - 1){
            resize(stackArraySize * 2);
        }
        stack[sizeVar++] = t;
    }

    @Override
    public T pop() {
        T t = (T) stack[sizeVar - 1];
        stack[sizeVar--] = null;
        return t;
    }


    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private int index = 0;
            @Override
            public boolean hasNext() {
                return index < sizeVar;
            }

            @Override
            public T next() {
                return (T) stack[index++];
            }
        };

    }

    @Override
    public void clear() {
        stack = new Object[4];
        sizeVar = 0;
    }

    @Override
    public int hashCode() {
        final int HASHCONST = 31;
        int result = 1;
        for (Object o : this) {
            result += result * HASHCONST + o.hashCode();
        }
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        return this.hashCode() == obj.hashCode() && obj.getClass() == this.getClass();
    }

    //PRIVATE METHODS
    private void resize(int newLength) {
        stackArraySize = newLength;
        Object[] newArray = new Object[newLength];
        System.arraycopy(stack, 0, newArray, 0, sizeVar);
        stack = newArray;
    }

}
